starlette/concurrency.py

```diff
@@ -1,43 +1,31 @@
-import asyncio
 import functools
-import sys
 import typing
 from typing import Any, AsyncGenerator, Iterator
 
+import anyio
+
 try:
     import contextvars  # Python 3.7+ only or via contextvars backport.
 except ImportError:  # pragma: no cover
     contextvars = None  # type: ignore
 
-if sys.version_info >= (3, 7):  # pragma: no cover
-    from asyncio import create_task
-else:  # pragma: no cover
-    from asyncio import ensure_future as create_task
 
 T = typing.TypeVar("T")
 
 
-async def run_until_first_complete(*args: typing.Tuple[typing.Callable, dict]) -> None:
```

Comment:
This is technically an API removal so I was wondering if anything used this. It looks like [some things do](https://sourcegraph.com/search?q=context:global+lang:Python+run_until_first_complete&patternType=literal), but not too many. FastAPI re-exports it but nobody seems to use that export.

Is there an easy anyio equivalent we can either implement here or put in the changelog? (I'm not suggesting updating the changelog in this PR, but I'll need to compile a changelog for 0.15 soon.)

Could we maybe keep this function for Starlette 0.15 but deprecate it? I don't think we really want this in our API because we don't use it much ourselves and it seems fairly rare that others do either. OTOH, the more stuff the next Starlette release breaks, the less people likely to use it.

Reply:
something like this?

```python
async def run_until_first_complete(*args: typing.Tuple[typing.Callable, dict]) -> None:
    async def run(handler, kwargs):
        await handler(**kwargs)
        tg.cancel_scope.cancel()

    async with anyio.create_task_group() as tg:
        for handler, kwargs in args:
            tg.start_soon(run, handler, kwargs)
```

Reply:
That's reasonable.  @graingert mentioned this too on gitter as well, so I'll add a compatible API back in.

@JayH5 how would we deprecate the API?  It's not documented, so would we show a `DeprecationWarning`?

Reply:
is it so bad to deprecated it? it's a handy function. Perhaps I'd prefer something matching trio_utils' wait_any and wait_all

Reply:
IMO a function like this is not documented and is an internal helper for concurrency.  It doesn't seem appropriate to me to expose it and support it as a first-class API of Starlette, long term.  Users of Starlette should look elsewhere for concurrency helpers.

Reply:
I've thought about it some more, and while I 100% agree with you, @uSpike, I think let's add it back but not deprecate it for now. That should leave this PR completely API-compatible except for the GraphQL change which is fine because that was old deprecated stuff and we're going to remove the whole thing anyway.

In some later version of Starlette we should go through the code more systematically and make APIs public or private. I think httpx went through this process at some point.

Reply:
(I guess the other API change is the testclient websocket context manager thing but there's a clear technical reason for that)

Reply:
@JayH5 added in 2b9dd22173e7ddb552e08503433c700601a417d0 with a test case since it is an api function used elsewhere.